File: flutter/lib/domain/repository/catalog_repository.dart
// flutter/lib/domain/repository/catalog_repository.dart
import 'package:grpc_demo/domain/model/price_update.dart';
import 'package:grpc_demo/domain/model/product.dart';

abstract interface class CatalogRepository {
  Future<Product> getProduct(String productId);
  Stream<Product> listProducts({String category = '', int page = 1});
  Stream<PriceUpdate> watchPrice(String productId);
}

File: android/app/src/main/kotlin/com/demo/grpcdemo/domain/repository/CatalogRepository.kt
package com.demo.grpcdemo.domain.repository

import com.demo.grpcdemo.domain.model.PriceUpdate
import com.demo.grpcdemo.domain.model.Product
import kotlinx.coroutines.flow.Flow

interface CatalogRepository {
    suspend fun getProduct(productId: String): Result<Product>
    fun listProducts(category: String = "", page: Int = 1): Flow<Product>
    fun watchPrice(productId: String): Flow<PriceUpdate>
}
